kuma/javascript/src/newsletter.jsx

```diff
@@ -89,97 +101,99 @@ export default function Newsletter() {
+                eventAction: 'progression',
+                eventLabel: 'error-forward'
+            });
+            return true;
+        }
 
-            fetch(newsletterSubscribeURL, {
-                method: 'POST',
-                headers: {
-                    'X-Requested-With': 'XMLHttpRequest',
-                    'Content-type': 'application/x-www-form-urlencoded'
-                },
-                body: params
-            }).then(response => {
-                response.json().then(responseJSON => {
-                    if (responseJSON.success !== 'success') {
-                        let errorsArray = responseJSON.errors;
+        event.preventDefault();
+        submitNewsletterSubscription(newsletterForm)
+            .then(({ success, errors }) => {
+                if (success) {
+                    permanentlyHideNewsletter();
+                    setShowSuccessfulSubscription(true);
+                    ga('send', {
+                        hitType: 'event',
+                        eventCategory: 'newsletter',
+                        eventAction: 'progression',
+                        eventLabel: 'complete'
+                    });
+                    return;
+                }
 
-                        // if there are erros and the array contain at least one item
-                        if (errorsArray && errorsArray.length) {
-                            setErrorsArray(errorsArray);
-                            // if there was an error, but there are no items in the array
-                        } else if (errorsArray && errorsArray.length === 0) {
-                            if (newsletterForm) {
-                                // set the skip-fetch data attribute on the form
-                                newsletterForm.dataset.skipFetch = 'true';
-                                // and submit again for diagnoses on the server
-                                newsletterForm.submit();
-                            }
-                        } else {
-                            setShowNewsletter(false);
-                            permanentlyHideNewsletter();
-                            setShowSuccessfulSubscription(true);
-                            ga('send', {
-                                hitType: 'event',
-                                eventCategory: 'newsletter',
-                                eventAction: 'progression',
-                                eventLabel: 'complete'
-                            });
-                        }
-                    }
-                });
+                if (errors && errors.length) {
+                    setError(errors);
+                } else if (errors && errors.length === 0) {
+                    // resubmit for diagnoses on the server (see skipFetch-
+                    // comment above)
+                    setSubmitToServer(true);
+                }
+            })
+            .catch(e => {
+                setError([e.toString()]);
```

Comment:
Ok. So I concede that this line would catch the error if `!response.ok`
https://codesandbox.io/embed/hardcore-dewdney-j6k98 
but it's just a bit weird that the result here will be about a problem with the JSON decoding. 

One neat compromise would be to change line 36 to:
```javascript
}).then(response => {
  if (!response.ok) {
    throw new Error(`Bad response from server: ${response.statusText}`);
  }
  return response.json()
});
```

Reply:
E.g. https://codesandbox.io/embed/peaceful-chatterjee-nwc7f

Either way, I'm now losing confidence that this is worth discussing. There are so many other flaws in Kuma other than the handling of some relatively rare error strings.